src/Analyzers/Core/Analyzers/UseObjectInitializer/AbstractUseObjectInitializerDiagnosticAnalyzer.cs

```diff
@@ -125,19 +126,22 @@ private void AnalyzeNode(SyntaxNodeAnalysisContext context)
                 var location1 = Location.Create(syntaxTree, TextSpan.FromBounds(
                     match.MemberAccessExpression.SpanStart, end));
 
-                RoslynDebug.AssertNotNull(UnnecessaryWithSuggestionDescriptor);
-                context.ReportDiagnostic(Diagnostic.Create(
-                    UnnecessaryWithSuggestionDescriptor, location1, additionalLocations: locations));
-
                 if (match.Statement.Span.End > match.Initializer.FullSpan.End)
                 {
-                    RoslynDebug.AssertNotNull(UnnecessaryWithoutSuggestionDescriptor);
-                    context.ReportDiagnostic(Diagnostic.Create(
-                        UnnecessaryWithoutSuggestionDescriptor,
-                        Location.Create(syntaxTree, TextSpan.FromBounds(
+                    context.ReportDiagnostic(DiagnosticHelper.CreateWithLocationTags(
+                        Descriptor,
+                        location1,
+                        ReportDiagnostic.Default,
+                        additionalLocations: locations.Add(Location.Create(syntaxTree, TextSpan.FromBounds(
                             match.Initializer.FullSpan.End,
-                            match.Statement.Span.End)),
-                        additionalLocations: locations));
+                            match.Statement.Span.End))),
+                        tagIndices: ImmutableDictionary<string, IEnumerable<int>>.Empty
+                            .Add(nameof(WellKnownDiagnosticTags.Unnecessary), new int[] { locations.Length })));
```

Comment:
I think it would be good to have another `CreateWithLocationTags` overload that takes an array of `additionalFadeLocations`. The helper would append to passed in non-faded `additionalLocations` and also compute the `tagIndices`.

Reply:
```csharp
public static Diagnostic CreateWithLocationTags(
            DiagnosticDescriptor descriptor,
            Location location,
            ReportDiagnostic effectiveSeverity,
            IEnumerable<Location> additionalNonFadingLocations,
            IEnumerable<Location> additionalFadingLocations,
            params object[] messageArgs)
```

Reply:
I think so too. I'll file a follow-up issue for this.

Reply:
i would like this.  i think it would make a lot of this cleaner and easier to understand (esp for people making changes or adding a new fading feture).

Reply:
➡️ Now implemented